### DISCUSSION TITLE
reporting internal telemetry from Tokio

### DISCUSSION THREAD

**Discussion by barkanido** [tokio-rs/tokio#3714] [Q&A]
Hey folks.
Can Tokio repost internal telemetry about itself? I am looking for stuff runtimes usually report, minus, of course, memory and GC which are obviously of less importance.

Maybe the number of tasks, scheduler-related telemetry, etc?

**Comment by MikailBag** [comment]
Tokio has `tracing` integration. AFAIK this way you can get some information about running and waiting-to-run tasks.

**Reply by barkanido** [reply]
thanks! can you point me to this API please? is it part of Tokio?

**Reply by MikailBag** [reply]
This integration is documented [here](https://docs.rs/tokio/1.5.0/tokio/#unstable-features).

AFAIU you should e.g. add `--cfg` `tokio_unstable` to RUSTFLAGS and enable `tracing` feature on tokio.

**Reply by barkanido** [reply]
hmm, so to my understanding this is not meant for production instrumentation?

**Reply by Darksonn** [reply]
Using tracing in production is perfectly fine. It requires the unstable flag because future updates of Tokio 1.x are not guaranteed to be backwards compatible when it comes to this feature — in particular, future versions of Tokio may require a newer version of tracing.

**Reply by barkanido** [reply]
I see! is there documentation of what becomes visible and how when this flag is set?

**Reply by Darksonn** [reply]
I'm pretty sure it doesn't add any new methods, but it will cause Tokio to emit a log message to tracing in various cases.

**Reply by davidbarsky** [reply]
> I'm pretty sure it doesn't add any new methods, but it will cause Tokio to emit a log message to tracing in various cases.

That is correct.
